tests/unit/test_minion.py

```diff
@@ -627,6 +627,61 @@ def test_minion_manage_beacons(self):
             self.assertIn("ps", minion.opts["beacons"])
             self.assertEqual(minion.opts["beacons"]["ps"], bdata)
 
+    def test_master_type_failover(self):
+        """
+        Tests master_type "failover" to not fall back to 127.0.0.1 address when master does not resolve in DNS
+        """
+
+        mock_opts = salt.config.DEFAULT_MINION_OPTS.copy()
+        mock_opts.update(
+            {"master_type": "failover", "master": ["master1", "master2"], "__role": ""}
+        )
+
+        class MockPubChannel:
+            def connect(self):
+                raise SaltClientError("MockedChannel")
+
+            def close(self):
+                return
+
+        def mock_resolve_dns(opts, fallback=False):
+            self.assertFalse(fallback)
+
+            if opts["master"] == "master1":
+                raise SaltClientError("Cannot resolve {0}".format(opts["master"]))
+
+            return {"master_ip": "192.168.2.1", "master_uri": "tcp://192.168.2.1:4505"}
+
+        def mock_transport_factory(opts, **kwargs):
+            self.assertEqual(opts["master"], "master2")
+            return MockPubChannel()
+
+        io_loop = salt.ext.tornado.ioloop.IOLoop()
```

Comment:
There is a [tornado testing package](https://www.tornadoweb.org/en/stable/testing.html) that makes things a way easier for testing async code.

Reply:
I am not sure it deserves it because it doesn't test any async functionality. io_loop here isn't needed here at all, so I removed it.

Reply:
@DmitryKuzmenko Are we Ok now?